interfaces/builtin/unity8.go

```diff
@@ -0,0 +1,190 @@
+       member=DispatchURL
+       peer=(name=com.canonical.URLDispatcher,label=###SLOT_SECURITY_TAGS###),
+
+  # This is needed when the app is already running and needs to be passed in
+  # a URL to open. This is most often used with content-hub providers and
+  # url-dispatcher, but is actually supported by Qt generally.
+  dbus (receive)
+       bus=session
+       path=/###PLUG_DBUS_APPIDS###
+       interface=org.freedesktop.Application
+       member=Open
+       peer=(label=###SLOT_SECURITY_TAGS###),
+
+  # Unity launcher (e.g. app counter, progress, alert)
+  dbus (receive, send)
+       bus=session
+       path=/com/canonical/Unity/Launcher/###PLUG_DBUS_APPIDS###
+       peer=(name=com.canonical.Unity.Launcher,label=###SLOT_SECURITY_TAGS###),
+
+  # Content Hub (pasteboard, file transfers)
+  dbus (receive, send)
+       bus=session
+       interface=com.ubuntu.content.dbus.Service
+       path=/
+       peer=(name=com.ubuntu.content.dbus.Service,label=###SLOT_SECURITY_TAGS###),
+
+  # Lttng tracing is very noisy and should not be allowed by confined apps.
+  # Can safely deny. LP: #1260491
+  deny /{,var/}{dev,run}/shm/lttng-ust-* r,
+`)
+
+var unity8ConnectedPlugSecComp = []byte(`
+recvfrom
+recvmsg
+sendmsg
+sendto
+shutdown
+`)
+
+type Unity8Interface struct{}
+
+func (iface *Unity8Interface) Name() string {
+	return "unity8"
+}
+
+func (iface *Unity8Interface) String() string {
+	return iface.Name()
+}
+
+func (iface *Unity8Interface) dbusAppId(app *snap.AppInfo) []byte {
+	var retval bytes.Buffer
+
+	appidbits := []string{app.Snap.Name(), app.Name, app.Snap.Revision.String()}
+	appid := strings.Join(appidbits, "_")
+
+	for _, value := range appid {
+		if unicode.In(rune(value), unicode.Letter, unicode.Digit) {
+			retval.WriteRune(rune(value))
+		} else {
+			retval.WriteString(fmt.Sprintf("_%2.2x", value))
+		}
+	}
+
+	return retval.Bytes()
+}
+
+func (iface *Unity8Interface) PermanentPlugSnippet(plug *interfaces.Plug, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	return nil, nil
+}
+
+func (iface *Unity8Interface) ConnectedPlugSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		oldTags := []byte("###SLOT_SECURITY_TAGS###")
+		newTags := slotAppLabelExpr(slot)
+		snippet := bytes.Replace(unity8ConnectedPlugAppArmor, oldTags, newTags, -1)
+
+		appidsOld := []byte("###PLUG_DBUS_APPIDS###")
+		var appidsNew bytes.Buffer
+		var appidsList []string
+		for _, app := range plug.Apps {
+			appidsList = append(appidsList, string(iface.dbusAppId(app)))
+		}
+		sort.Strings(appidsList) // makes tests reliable
+		if len(appidsList) == 1 {
+			appidsNew.WriteString(appidsList[0])
+		} else if len(appidsList) > 1 {
+			appidsNew.WriteByte('{')
+			appidsNew.WriteString(strings.Join(appidsList, ","))
+			appidsNew.WriteByte('}')
+		}
+		snippet = bytes.Replace(snippet, appidsOld, appidsNew.Bytes(), -1)
```

Comment:
I understand that the Snapd AppID is different than the Unity8 AppID, but I don't think that we should require Unity8 to change to be able to run on snapd. There are likely other desktop environments that have their own application identifiers that we should honor and allow them to use as well. So while we could discuss changing the AppID in Unity8 (unfortunately this would probably be a hard change as there's no shared implementation) I don't think that it should restrict whether different application identifiers are allowed by projects using snapd.

I also think that for Unity8 to move to snapd AppIDs we'll need them to include the revision. That way we can properly handle running applications while they're upgraded, removed or installed. But that is a *much* larger conversation that this PR.

As discussed on IRC, think that a way forward is to use a glob instead of specifying the AppID format today. It will provide less specificity in the short term, but this interface is also restricted currently, so there won't be many users. That way we can move forward with getting apps and desktops moved over to using snapd confined.